core/src/saros/concurrent/management/HeartbeatDispatcher.java

```diff
@@ -1,72 +1,106 @@
+
+  private static final Logger log = Logger.getLogger(HeartbeatDispatcher.class);
+
   private final ISarosSession session;
   private final UISynchronizer uiSynchronizer;
-  private final JupiterClient jupiter;
+  private final JupiterClient jupiterClient;
+
+  private ScheduledThreadPoolExecutor heartbeatScheduledExecutor;
 
   public HeartbeatDispatcher(
-      ISarosSession sarosSession,
-      UISynchronizer uiSynchronizer,
-      ConcurrentDocumentClient cdClient) {
-    this.session = sarosSession;
+      final ISarosSession session,
+      final UISynchronizer uiSynchronizer,
+      final ConcurrentDocumentClient documentClient) {
+
+    this.session = session;
     this.uiSynchronizer = uiSynchronizer;
-    this.jupiter = cdClient.getJupiterClient();
+    this.jupiterClient = documentClient.getJupiterClient();
   }
 
   @Override
   public void start() {
+
     session.addActivityProducer(this);
+
     heartbeatScheduledExecutor =
         new ScheduledThreadPoolExecutor(1, new NamedThreadFactory("JupiterHeartbeat"));
-    // Schedule the heartbeat once per minute, this should be absolutely
-    // sufficient.
-    // NOTE: This might even be to frequent for a lot of open documents, but
-    // that likely
-    // causes other performance problems anyway.
+
+    /* Schedule the heart-beat once per minute, this should be absolutely sufficient.
+     * NOTE: This might even be to frequent for a lot of open documents, but
+     * that likely causes other performance problems anyway.
+     * Suggestion: store a map of the documents. Invoke the heart-beat more frequently but only fire heart-beats for a subset of documents so
+     * that we do not produce too much traffic at once.
+     * E.G: We have 50 documents, on the first interval fire changes for the first 10 documents.
+     * On the next interval fire changes for the next 10 documents on so on.
+     */
+
+    // client documents should only be accessed by the main thread
     heartbeatScheduledExecutor.scheduleWithFixedDelay(
-        this::dispatchHeartbeats, 1, 1, TimeUnit.MINUTES);
+        () -> uiSynchronizer.syncExec(this::dispatchHeartbeats), 1, 1, TimeUnit.MINUTES);
   }
 
   private void dispatchHeartbeats() {
-    // clientDocs should only be accessed by the main thread
-    // FIXME syncExec not asyncExec
-    uiSynchronizer.asyncExec(
-        () ->
-            jupiter
-                .getClientDocs()
-                .forEach(
-                    (path, jupiter) -> {
-                      JupiterActivity heartbeat =
-                          jupiter.generateJupiterActivity(
-                              new NoOperation(), session.getLocalUser(), path);
-                      fireActivity(heartbeat);
-                    }));
+
+    assert uiSynchronizer.isUIThread() : "invalid thread access";
+
+    final User localUser = session.getLocalUser();
+
+    for (Entry<SPath, Jupiter> entry : jupiterClient.getClientDocs().entrySet()) {
```

Comment:
Why expand the `forEach` call into a verbose for loop?
This is much harder to read and I would strongly vote for the brevity of the first approach.

I cannot spot any functional differences here, except for moving the `uiSynchronizer`, so why even bother?

Reply:
In general, I would agree with this. Especially when iterating over maps, lambdas are preferable in my mind as they allow you to directly access both key and value without the usage of a map entry.

In this case, the following would be the shortest version:
```java
    jupiterClients
        .getClientDocs()
        .forEach(
            (path, jupiterAlgorithm) ->
                fireActivity(
                    jupiterAlgorithm.generateJupiterActivity(
                        new NoOperation(), sarosSession.getLocalUser(), path)));
```
<details><summary>(In this case, explicitly declaring the activity first doesn't really improve the formatting.)</summary>

```java
    jupiterClients
        .getClientDocs()
        .forEach(
            (path, jupiterAlgorithm) -> {
              JupiterActivity heartbeat =
                  jupiterAlgorithm.generateJupiterActivity(
                      new NoOperation(), sarosSession.getLocalUser(), path);

              fireActivity(heartbeat);
            });
```

</details>

In this case, the indentation is not really optimal, so the readability is down to personal preference in my mind (I think I would use the lambda) . I am fine with both options (lambda or loop).

Reply:
Well the Google Formatter can format things in not such a nice way and I do not like it when the code starts to "drawing arrows".

On the other hand I really do not like it if Lambas (which should represent the art of functional programming) calling non idempotent methods.

Beside that the following call would be more of a functional style:

    this.jupiter
        .getClientDocs()
        .entrySet()
        .stream()
        .map(
            (entry) ->
                entry
                    .getValue()
                    .generateJupiterActivity(
                        new NoOperation(), session.getLocalUser(), entry.getKey()))
        .forEach((activity) -> fireActivity(activity));